File: C++/Save.cpp
#include <fstream>
#include <string>
#include <stdexcept>
#include <ctime>

using namespace std;

class Logger {
public:
    explicit Logger(const string& caminho) {
        arquivo_.open(caminho, ios::app);
        if (!arquivo_.is_open())
            throw runtime_error("Não foi possível abrir: " + caminho);

        arquivo_ << "[" << timestamp() << "] === Sessão iniciada ===\n";
        arquivo_.flush();
    }

    Logger(const Logger&)            = delete;
    Logger& operator=(const Logger&) = delete;

    void registrar(const string& mensagem) {
        arquivo_ << "[" << timestamp() << "] " << mensagem << "\n";
        arquivo_.flush();
    }

    ~Logger() {
        if (arquivo_.is_open()) {
            arquivo_ << "[" << timestamp() << "] === Sessão encerrada ===\n";
            arquivo_.close();   
        }
    }

private:
    ofstream arquivo_;

    static string timestamp() {
        time_t agora = time(nullptr);
        char buf[20];
        strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", localtime(&agora));
        return buf;
    }
};


int main() {
    Logger logger("sessao.log");          
    logger.registrar("Aplicação iniciada.");
    logger.registrar("Processando dados...");

    return 0;
}

File: python/Save.py
import datetime

class Logger:
    def __init__(self, caminho_arquivo: str):
        self._arquivo = open(caminho_arquivo, "a", encoding="utf-8")
        inicio = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self._arquivo.write(f"[{inicio}] === Sessão iniciada ===\n")
        self._arquivo.flush()

    def registrar(self, mensagem: str):
        agora = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self._arquivo.write(f"[{agora}] {mensagem}\n")
        self._arquivo.flush()

    def __del__(self):
        if self._arquivo and not self._arquivo.closed:
            fim = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            self._arquivo.write(f"[{fim}] === Sessão encerrada ===\n")
            self._arquivo.close()

if __name__ == "__main__":
    logger = Logger("sessao.log")
    logger.registrar("Aplicação iniciada.")
    logger.registrar("Processando dados...")
   